content/en/blog/_posts/2023-03-08-memory-qos-cgroups-v2/index.md

```diff
@@ -0,0 +1,275 @@
+  
+  The memory limit specifies the memory limit, beyond which if the container tries to allocate 
+  more memory, Linux kernel will terminate it with an OOM (Out of Memory) kill.
+
+  In cgroup v1, `memory.limit_in_bytes` interface is used to set the memory usage limit.
+  However, unlike CPU, it was not possible to apply memory throttling: as soon as a
+  container crossed the memory limit, it would be OOM killed.
+
+  In cgroup v2, `memory.max` is analogous to `memory.limit_in_bytes` in cgroupv1.
+  Memory QoS maps `memory.max` to `spec.containers[].resources.limits.memory` to
+  specify the hard limit for memory usage. If the memory consumption goes above this
+  level, the kernel invokes its OOM Killer.
+
+  cgroup v2 also added `memory.high` configuration . Memory QoS uses `memory.high`
+  to set memory usage throttle limit. If the `memory.high` limit is breached,
+  the offending cgroups are throttled, and the kernel tries to reclaim memory
+  which may avoid an OOM kill.
+
+## How it works
+
+### Cgroup v2 memory controller interfaces & Kubernetes container resources mapping
+
+Memory QoS uses the memory controller of cgroup v2 to guarantee memory resources in
+Kubernetes. cgroupv2 interfaces that this feature: 
+* `memory.max`
+* `memory.min`
+* `memory.high`.
+
+![](./memory-qos-cal.svg)
+
+`memory.max` is mapped to memory limit specified in the Pod spec. The kubelet (and 
+the container runtime) enforces the limit and prevent the container from using more
+than the configured resource limit. If a process in a container tries to consume
+more than the specified limit, kernel terminates the process with an out of
+memory Out of Memory (OOM) error. 
+
+
+![](./container-memory-max.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+`memory.min` is mapped to memory requests, which results in reservation of memory resources
+that can never reclaimed by the kernel. This is how Memory QoS ensures the availability of 
+memory for Kubernetes pods. If there's no unprotected reclaimable memory available, OOM 
+killer is invoked.
+
+![](./container-memory-min.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+For memory protection, in addition to the original way of limiting memory usage, Memory QoS
+throttles workload approaching its memory limit, ensuring that the system is not overwhelmed
+by sporadic increases in memory usage. A new field, MemoryThrottlingFactor, is introduced in
+the KubeletConfiguration which is set to 0.9 by default. `memory.high` is mapped to throttling
+limit calculated by using MemoryThrottlingFactor, `spec.containers[].resources.requests.memory`
+and `spec.containers[].resources.limits.memory` as in the formula below, and rounding down the
+value to the nearest page size:
+
+![](./container-memory-high.svg)  
+<sub>i: the i<sup>th</sup> container in one pod</sub>
+
+**Note**: If a container has no memory limit specified, we substitute limits.memory for node allocatable memory.
```

Comment:
Try one of these:
```suggestion
**Note**: If a container has no memory limit specified, Kubernetes substitutes `limits.memory` for node allocatable memory.
```
or
```suggestion
**Note**: If a container has no memory limit specified, substitute `limits.memory` for node allocatable memory.
```
In the second one “substitute” is short for “you should substitute”.